src/openrct2/ride/Vehicle.cpp

```diff
@@ -1486,6 +1486,10 @@ bool Vehicle::OpenRestraints()
         if (curRide == nullptr)
             continue;
 
+        int currentVehicleIndex = vehicle->vehicle_type;
+        while (vehicle->ForceShiftDown() && vehicle->vehicle_type != currentVehicleIndex)
```

Comment:
This could alternatively be tracking how many shifts it performs and stopping after a certain number, because it is possible to create an infinite loop.